core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -310,126 +485,361 @@ private void initialize() {
+        // the respective records.
+        private InFlightState batchState;
+
+        // The offset state map is used to track the state of the records per offset. However, the
+        // offset state map is only required when the state of the offsets within same batch are
+        // different. The states can be different when explicit offset acknowledgment is done which
+        // is different from the batch state.
+        private NavigableMap<Long, InFlightState> offsetState;
+
+        InFlightBatch(String memberId, long firstOffset, long lastOffset, RecordState state,
+            int deliveryCount, AcquisitionLockTimerTask acquisitionLockTimeoutTask
+        ) {
+            this.firstOffset = firstOffset;
+            this.lastOffset = lastOffset;
+            this.batchState = new InFlightState(state, deliveryCount, memberId, acquisitionLockTimeoutTask);
         }
 
-        @Override
-        public int hashCode() {
-            return Objects.hash(state, deliveryCount, memberId);
+        // Visible for testing.
+        long firstOffset() {
+            return firstOffset;
         }
 
-        @Override
-        public boolean equals(Object o) {
-            if (this == o) {
-                return true;
+        // Visible for testing.
+        long lastOffset() {
+            return lastOffset;
+        }
+
+        // Visible for testing.
+        RecordState batchState() {
+            if (batchState == null) {
+                throw new IllegalStateException("The batch state is not available as the offset state is maintained");
+            }
+            return batchState.state;
+        }
+
+        // Visible for testing.
+        String batchMemberId() {
+            if (batchState == null) {
+                throw new IllegalStateException("The batch member id is not available as the offset state is maintained");
+            }
+            return batchState.memberId;
+        }
+
+        // Visible for testing.
+        int batchDeliveryCount() {
+            if (batchState == null) {
+                throw new IllegalStateException("The batch delivery count is not available as the offset state is maintained");
+            }
+            return batchState.deliveryCount;
+        }
+
+        // Visible for testing.
+        NavigableMap<Long, InFlightState> offsetState() {
+            return offsetState;
+        }
+
+        private InFlightState tryUpdateBatchState(RecordState newState, boolean incrementDeliveryCount, int maxDeliveryCount, String newMemberId) {
+            if (batchState == null) {
+                throw new IllegalStateException("The batch state update is not available as the offset state is maintained");
+            }
+            return batchState.tryUpdateState(newState, incrementDeliveryCount, maxDeliveryCount, newMemberId);
+        }
+
+        private void maybeInitializeOffsetStateUpdate() {
```

Comment:
Can we have some javadoc for this function

Reply:
Not neccessarily needed for a private internal method. I have skipped writing details for it.